Get Implementation in AsyncCache that takes in a callback function

sphill99 opened the pull request:
Implemented the get function in AsyncCache that takes in a callback function in two different ways. In the one (DiskBasedAsyncCache), it uses an asynchronous file channel, and for the fallback version, it works very similarly to the current DiskBasedCache. Most of the code in those files are from that, aside from the get method. There are not tests yet for this, as not enough has really been implemented to test it, but was more to check the direction of the code.

jbir789 commented:
Hi Scott, I'd recommend to update the PR title to be a bit more descriptive. I find that the first sentence of your description provides much more context on what this PR is about, so feel free to use it as the title.

Descriptive titles are important is it allows to quickly get a sense of what the PR is about.

jbir789 commented on src/main/java/com/android/volley/AsyncCache.java:
```diff
@@ -0,0 +1,82 @@
+        final AtomicReference<Entry> entryRef = new AtomicReference<>();
+        get(
+                key,
+                new OnGetCompleteCallback() {
```
nit: we could use a lambda expression here for simplicity (I believe [all version of Android can now use a lambda](https://developer.android.com/studio/write/java8-support) when using a recent Android Gradle plugin)

This would convert your code into:

```java
get(
    key, 
    entry -> {
      entryRef.set(entry);
      latch.countDown();
    }
);
```

For context, lambdas are like a more concise way to write an anonymous class. Feel free to take a look at the official java doc about lambdas if it's the first time you hear about them / use them: https://docs.oracle.com/javase/tutorial/java/javaOO/lambdaexpressions.html

sphill99 replied:
When I tried to make that change, it says lambda expressions aren't allowed in Java 7. Even though I thought Volley uses java8. I'll look into this.

jpd236 replied:
Volley actually is built with minSdkVersion = 8, which is actually Java 6 :)

Lambdas have been backported in principle per https://developer.android.com/studio/write/java8-support - not sure offhand how far back this goes. But I'm not 100% sure what the story is offhand in terms of exposing a library that relies on this support.

While it's not ideal, unless we can pretty quickly and confidently find an answer suggesting we can safely use lambdas without introducing major complexity for apps depending on the library using this desugaring support, we should probably avoid them except when writing code that is guaranteed to run on the right minSdkVersion.

Note that this also includes try-with-resources.

jpd236 replied:
FWIW, I'm open to considering increasing the minSdkVersion somewhat. At this point it's exceedingly rare to target anything under API 14, which is what the Android support library requires; that seems like a reasonable point to increase it, if that'd be helpful. But it'd be nice to have a clear benefit to doing so, and 14 won't get us native try-with-resources or lambdas AFAIK.

joebowbeer replied:
To be clear, the current Java target will support try-with-resources if sdk >= 19 

```
compileOptions {
    sourceCompatibility JavaVersion.VERSION_1_7
    targetCompatibility JavaVersion.VERSION_1_7
}
```

Though I think try-with-resources can obscure what's going on, esp. in nested and other complex situations, so no loss of any significance.

https://stackoverflow.com/a/19470405/901597

Java 8 is needed for lambda.

jbir789 commented on src/main/java/com/android/volley/AsyncCache.java:
```diff
@@ -0,0 +1,82 @@
+    // TODO: Handle callback never being invoked? Match behavior when sync cache throws
+    // RuntimeException.
+
+    @Override
```
Could you also annotate the method as `@Nullable`? It acts a documentation to tell the caller that the return value could be null.

sphill99 replied:
fixed

jbir789 commented on src/main/java/com/android/volley/AsyncCache.java:
```diff
@@ -0,0 +1,82 @@
+public abstract class AsyncCache implements Cache {
+
+    public interface OnGetCompleteCallback {
+        void onGetComplete(Entry entry);
```
Since `entry` can be null, please annotate the parameter as `@Nullable`, so whoever implements the class will be aware that the method parameter can be null.

sphill99 replied:
fixed

jbir789 commented on src/main/java/com/android/volley/AsyncCache.java:
```diff
@@ -0,0 +1,82 @@
+        put(
+                key,
+                entry,
+                new OnPutCompleteCallback() {
```
You can use a lambda here as well.

jbir789 commented on src/main/java/com/android/volley/AsyncCache.java:
```diff
@@ -0,0 +1,82 @@
+        }
+    }
+
+    // TODO: Implement the rest.
```
nit: did you mean to move this below `put`?

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,249 @@
+import java.util.Map;
+
+@SuppressLint("NewApi")
+public class DiskBasedAsyncCache extends AsyncCache {
```
As for `AsyncCache`, make sure to add javadoc to describe what the class does and how to use it.

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,249 @@
+    /** Clears the cache. Deletes all cached files from disk. */
+    @Override
+    public synchronized void clear() {
+        // TODO (sphill99): Implement
```
nit: remove the space between `TODO` and `(sphil99)`

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,249 @@
+    /** Returns the cache entry with the specified key if it exists, null otherwise. */
+    @Override
+    public void get(String key, OnGetCompleteCallback callback) {
+        final OnGetCompleteCallback cb = callback;
```
nit: prefer spelling out variable names so it's more clear to the reader. You could call it `getCallback`?

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,249 @@
+        File file = getFileForKey(key);
+        Path path = Paths.get(file.getPath());
+        try {
+            AsynchronousFileChannel afc =
```
This is a closable resource, which means it needs to be closed after being used.

A nice way to achieve this is to use a [try-with-resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html):

```java
try (AsynchronousFileChannel fileChannel = AsynchronousFileChannel.open(path, StandardOpenOption.READ) {
  // do your work here
} catch (Exception e) {
  // handle exception
}
```

The try-with-resources will automatically close the resource when it completes or when an exception is thrown.

jbir789 replied:
Since we can't use try-with-resources. make sure to close the file channel when the try/catch block is completed

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,249 @@
+                            cb.onGetComplete(null);
+                        }
+                    });
+        } catch (Exception e) {
```
The `DiskBasedCache` only catchtes `IOException`s. Is there a reason why we want to catch any `Exception`s in here?

sphill99 replied:
Oh no, I think I just put that there when I first wrote the try/catch, and never changed it. I'll fix that now.

sophiahaim commented on src/main/java/com/android/volley/AsyncCache.java:
```diff
@@ -0,0 +1,82 @@
+
+    public abstract void get(String key, OnGetCompleteCallback callback);
+
+    // TODO: Handle callback never being invoked? Match behavior when sync cache throws
```
nit: Add the issue in parentheses:
// TODO(#181): Handle...

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,249 @@
+
+    /** Handles holding onto the cache headers for an entry. */
+    @VisibleForTesting
+    static class CacheHeader {
```
Could `CacheHeader` be shared between this class and `DiskBaseCache`? The implementations don't look the same at this time, but just wondering what's the long term plan and if it would make sense to extract the class to a top-level class so it can be shared.

sphill99 replied:
Yeah, I was thinking about making another abstract class that these two extend. This would allow us to not repeat the shared methods/class. Currently, I think the write and read methods would be implemented differently, so I don't know how that'd work. But I think that's definitely better to try to do that.

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCacheFallback.java:
```diff
@@ -0,0 +1,454 @@
+import java.util.List;
+import java.util.Map;
+
+public class DiskBasedAsyncCacheFallback extends AsyncCache {
```
Could you also add doc here to explain how the class differs from `DiskBasedAsyncCache`?

sphill99 replied:
added

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCacheFallback.java:
```diff
@@ -0,0 +1,454 @@
+        File file = getFileForKey(key);
+
+        try {
+            DiskBasedAsyncCacheFallback.CountingInputStream cis =
```
I believe you can use a try-with-resources here as well.

sphill99 replied:
It says that it requires an API of 19 or higher to do a try-with-resources.

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCacheFallback.java:
```diff
@@ -0,0 +1,454 @@
+        File file = getFileForKey(key);
+
+        try {
+            DiskBasedAsyncCacheFallback.CountingInputStream cis =
```
nit: I think you may not need to add the `DiskBaseAsyncCacheFallback` prefix since it's an inner class of this class. (same below)

sphill99 replied:
yep, just fixed this.

sophiahaim commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,249 @@
+    @Override
+    public void get(String key, OnGetCompleteCallback callback) {
+        final OnGetCompleteCallback cb = callback;
+        final DiskBasedAsyncCache.CacheHeader entry = mEntries.get(key);
```
Check that entry is non-null

sphill99 replied:
fixed

sophiahaim commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,249 @@
+            afc.read(
+                    buffer,
+                    0,
+                    buffer,
```
Add args to these two to clarify what they are:
/* position= */ 0,
/* attachment= */ buffer

sphill99 replied:
added those

sophiahaim commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,249 @@
+        // TODO (sphill99): Implement
+    }
+
+    /**
```
I think some of these common functions that are the same between your new Cache implementations and the existing one could be factored out into a Util that all three classes could call

You may be able to do the same with some of the CONSTANTS at the top of the file that seem to be the same for all three

sphill99 replied:
Yeah, I want to extract some of the stuff that's repeated out to a higher level class. Should I do that for this PR?

joebowbeer commented:
What is test coverage for this new code? I don't see any added tests.

sphill99 commented:
@joebowbeer I still haven't written tests for this function. I'm going to write tests for the AsyncCache as a whole, including this implementation. I just wanted to get the general structure reviewed before I move further with the implementations. Currently these implementations aren't being used.

I'll make sure to keep the existing coverage.

sophiahaim commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,145 @@
+                    AsynchronousFileChannel.open(path, StandardOpenOption.READ);
+            ByteBuffer buffer = ByteBuffer.allocate((int) file.length());
+            afc.read(
+                    /* destination */ buffer,
```
nit: add "="
e.g. /* destination= */ buffer

sphill99 replied:
fixed

sophiahaim commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCacheFallback.java:
```diff
@@ -0,0 +1,152 @@
+    /** The maximum size of the cache in bytes. */
+    private final int mMaxCacheSizeInBytes;
+
+    /** Default maximum disk usage in bytes. */
```
Are these used anywhere yet? If not, I think hold off on including them until they're used.

When you do need to add them, I think you can put them in your new Utility class since they are the same for both implementations (async and fallback)

sphill99 replied:
I removed these for now

sophiahaim commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCacheFallback.java:
```diff
@@ -0,0 +1,152 @@
+
+    /** Returns the cache entry with the specified key if it exists, null otherwise. */
+    @Override
+    public synchronized void get(String key, OnGetCompleteCallback callback) {
```
This is really similar to the existing DiskBasedCache, except instead of returning null or entry.toCacheEntry, you call callback.onGetComplete(null/entry.toCacheEntry)

I'm wondering if it would be possible to use the existing DiskBasedCache in here and just call #get, then do callback.onGetComplete with whatever it returns. There may be a couple ways of approaching this, so happy to discuss more offline too.

I think it would generally simplify this class if you can leverage the existing implementation.

jpd236 replied:
+1 - for the fallback, we should try to share as much code with possible as the existing implementation rather than duplicating it.

In fact, though, I wonder if we really want to have a second class at all. The reality is that it's confusing to have a class extend AsyncCache but not actually be Async - how should the user of this object determine whether the cache supports true Async operation (and thus doesn't need to be called on a blocking thread) or not (and thus does)?

The class which uses this will ultimately need to make a decision based on the capabilities of the cache to either use the fully async API or else send the task to a blocking thread pool. But in the latter case, there's actually no reason to use the async API methods, since they're blocking anyway.

This might become clearer when we actually implement the async cache handling step, but my instinct here is that we actually don't need this class, and should instead have the dispatcher accept either a vanilla Cache (which defaults to DiskBasedCache) or an AsyncCache (which defaults to DiskBasedAsyncCache, where supported). If the cache is an instance of AsyncCache, it knows it can safely use the async methods. If not, it should use the fallback and schedule the read for a blocking background thread.

sphill99 replied:
I changed it so the fallback now doesn't repeat the code. But yeah, that makes sense.

jpd236 replied:
That's certainly better, but then why would we need it at all per my comment? Would it make sense to hold off on adding it unless/until we see how it would be helpful/needed, or do you see a reason that the next level up would make use of this, given that it will need to know how to dispatch the cache operations?

sphill99 replied:
oh okay, I understand what you're saying now. I just deleted it in the last commit.

jpd236 commented on src/main/java/com/android/volley/AsyncCache.java:
```diff
@@ -0,0 +1,96 @@
+
+    @Nullable
+    @Override
+    public Entry get(String key) {
```
I think we should mark the legacy methods as final - subclasses of AsyncClass should not be modifying these implementations, only implementing the async versions. (At least until someone can provide a valid use case for it; until then, the tighter we keep the API, the stronger the guarantees the implementation can offer).

sphill99 replied:
fixed

jpd236 commented on src/main/java/com/android/volley/AsyncCache.java:
```diff
@@ -0,0 +1,96 @@
+        try {
+            latch.await();
+            return entryRef.get();
+        } catch (InterruptedException e) {
```
We should probably log the exception here as well (with VolleyLog). Otherwise, the operation might not have gone through, but apart from the thread's interrupt status (which is often unchecked by callers), there'd be no visible indication that something had been canceled.

joebowbeer replied:
Consider awaitUninteruptibly. Is this thread exposed enough to be interrupted? If not, this handling is just another section of untested code.

jpd236 replied:
awaitInterruptibly isn't a method on CountDownLatch.

While I agree interrupt handling is a mess in Java, I think we can at least ensure we handle them correctly to the extent possible within Volley (we do have RequestQueue#stop which would stop the executors if we control them).

joebowbeer replied:
At least a Future (incl. CompletableFuture) can be canceled.

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,145 @@
+    private final int mMaxCacheSizeInBytes;
+
+    /** Default maximum disk usage in bytes. */
+    private static final int DEFAULT_DISK_USAGE_BYTES = 5 * 1024 * 1024;
```
can we share magic constants like this with DiskBasedCache at least until we need to fork them? e.g. by putting them in DiskBasedCacheUtility

sphill99 replied:
fixed

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,145 @@
+                        @Override
+                        public void completed(Integer result, ByteBuffer attachment) {
+                            if (attachment.hasArray()) {
+                                int offset = attachment.arrayOffset();
```
nit: looks like this variable is unused

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,145 @@
+                    /* attachment */ buffer,
+                    new CompletionHandler<Integer, ByteBuffer>() {
+                        @Override
+                        public void completed(Integer result, ByteBuffer attachment) {
```
wondering whether for safety, we need to read the value of result and truncate the entry to that size (/ handle if it is -1, per the Javadoc).

With the old cache, we'd occasionally see strange bugs due to files being corrupted, possibly due to parallel access. While that's a misuse of Volley if so, it is good if we can gracefully fall back to the extent possible. And it's possible the size changes between when we read the length / allocate the buffer and when we complete the actual data read. So the buffer might have garbage data at the end (if the file was truncated), or might be incomplete (if it grew). In both cases we should probably assume the entry is invalid and avoid returning it.

sphill99 replied:
I added a check to see if the file length is the same as the size allocated before the read.

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,145 @@
+                        }
+
+                        @Override
+                        public void failed(Throwable exc, ByteBuffer attachment) {
```
log the exception so the failure is visible

sphill99 replied:
fixed

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,145 @@
+                        }
+                    });
+        } catch (IOException e) {
+            cb.onGetComplete(null);
```
log exception

sphill99 replied:
fixed

jpd236 replied:
better, but see above note regarding how to handle errors

joebowbeerxealth replied:
Line 107 is still potentially a second call to the callback

joebowbeer replied:
Line 107 is still potentially a second invocation of callback.

See comment above and this example of another way to structure this:

https://github.com/google/volley/blob/master/src/main/java/com/android/volley/toolbox/DiskBasedCache.java#L148

sphill99 commented:
I should have made appropriate edits for all the PR comments, or left questions on them. It is ready to be reviewed again.

jbir789 commented on src/main/java/com/android/volley/toolbox/CacheHeader.java:
```diff
@@ -0,0 +1,147 @@
+/** Handles holding onto the cache headers for an entry. */
+@VisibleForTesting
+class CacheHeader {
+    /**
+     * The size of the data identified by this CacheHeader on disk (both header and data).
+     *
+     * <p>Must be set by the caller after it has been calculated.
+     *
+     * <p>This is not serialized to disk.
+     */
```
This comment looks intended for `long size;` below

jbir789 commented on src/main/java/com/android/volley/toolbox/CacheHeader.java:
```diff
@@ -0,0 +1,147 @@
+    final String key;
+
+    /** ETag for cache coherence. */
+    final String etag;
```
nit: please annotate as `@Nullable`

sphill99 replied:
fixed

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,136 @@
+                        public void completed(Integer result, ByteBuffer attachment) {
+                            // if the file size changes, return null
+                            if (size != file.length()) {
+                                cb.onGetComplete(null);
```
maybe log a warning in this case since it implies the file changed while we were reading it.

sphill99 replied:
added

joebowbeer replied:
Can this mean that the read terminated early? I would treat this as a failure.

sphill99 replied:
I think it means the file changed between the time we opened the channel and finished the reading. The size allocated and the file length shouldn't be affected by the success of the read. I think if the read fails, it would never reach the completed block.

joebowbeer replied:
In either case, it sounds like an error to me. I don't think it serves much purpose to check the file length if it will be ignored for all practical purposes. I would either ignore the file length and use the bytes read, if this is reliable, or use the file length as a correctness check, if not.

jpd236 replied:
I think the confusion here is - this should say "if (size != result)". We shouldn't compare the file size before and after - we should compare the size of the file when we allocated the buffer to the number of bytes we actually read.

sphill99 replied:
but the number of bytes we actually read is dependent upon the size of the buffer we allocate. This would work in the case of the read bytes being less than the allocated, but it will never be more.

jpd236 replied:
true. but we should still look at the number of bytes read and make sure it's not less, which could happen even if the file size is now the same.

I think checking for result != size || current file size != size would be reasonable. Or, just the former, and let the latter be handled by integrity checks on the cache header.

sphill99 replied:
Okay, I'll make that fix

sphill99 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,148 @@
+                    afc.close();
+                } catch (IOException e) {
+                    VolleyLog.d("%s: %s", file.getAbsolutePath(), e.toString());
+                    cb.onGetComplete(null);
```
oh thanks, just fixed

sphill99 commented:
I think I addressed all PR comments. It is ready to be reviewed again.

jpd236 commented on src/main/java/com/android/volley/AsyncCache.java:
```diff
@@ -0,0 +1,97 @@
+                new OnPutCompleteCallback() {
+                    @Override
+                    public void onPutComplete(boolean success) {
+                        if (success) {
```
AFAICT we need to count down the latch regardless of whether the put succeeds or fails. If we don't, the caller will just be blocking indefinitely even though nothing is still happening.

put doesn't provide a way to return an error, so we should make sure it's logged somewhere in the stack, but otherwise we should just return control to the caller once the put is done.

sphill99 replied:
would it be logged in this method, or in the async implementation, where we know what happened?

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,147 @@
+
+/**
+ * AsyncCache implementation that uses Java NIO's AsynchronousFileChannel to perform asynchronous
+ * disk reads and writes. This should only be used by devices with an API level of 26 or above.
```
I think you can remove this sentence and the `@SuppressLint` annotation and just add a `@RequiresApi(Build.VERSION_CODES.O)` instead. This is preferable because:

-  Instead of suppressing all lint checks, it only disables them for APIs added up through O. If you make an API call to something added in P, it will still fail, which is good as we would need additional checks.
-  It will enforce that the caller checks for O+ before using the class (or that it has a RequiresApi annotation of its own)
-  Its self-documenting - no need to repeat it in the Javadoc.

sphill99 replied:
Fixed this

jpd236 commented on src/main/java/com/android/volley/toolbox/CacheHeader.java:
```diff
@@ -0,0 +1,148 @@
+import java.util.List;
+
+/** Handles holding onto the cache headers for an entry. */
+@VisibleForTesting
```
this annotation doesn't make sense here; this class needs to be package-visible in order for other classes in this package to use it.

We use this annotation when we increase the visibility beyond what it otherwise would be for the _sole_ purpose of making it accessible in a unit test.

You should see Android Studio complaining about this, e.g. where DiskBasedAsyncCache is calling:

`cb.onGetComplete(entry.toCacheEntry(data));`

because it thinks toCacheEntry is only visible for testing, but here we're using it from prod code.

sphill99 replied:
fixed

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,147 @@
+                            public void completed(Integer result, ByteBuffer attachment) {
+                                // if the file size changes, return null
+                                if (size != file.length()) {
+                                    VolleyLog.d(
```
logging stuff:

1) use e instead of d - this is an error, not debugging info
2) You can do `VolleyLog.e("File changed while reading: %s", file.getAbsolutePath())` - no need to use format args for a constant string. Also, I think it's clearer to have the error be the first thing, and specific details like the file name come second.
3) Here and throughout, it should be "%s", not "s%", no?

jpd236 commented on src/main/java/com/android/volley/AsyncCache.java:
```diff
@@ -0,0 +1,97 @@
+                        if (success) {
+                            latch.countDown();
+                        } else {
+                            return;
```
no need for "return" statement as the last statement in a void method, here and below.

AS highlights these in yellow for me - do you not see that too? In general, we should aim to have no warnings in new code.

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,147 @@
+                                    return;
+                                }
+                                if (attachment.hasArray()) {
+                                    final int offset = attachment.arrayOffset();
```
what's the exact error message you get when you remove this line?

I think it's probably an ErrorProne thing which should probably be suppressed by adding a suppression for that, not by introducing a new variable that is itself unused and a compiler warning. arrayOffset() does nothing internally; we know there is a backing array because we allocated with ByteBuffer.allocate() above.

sphill99 replied:
The error is: "ByteBuffer.array() shouldn't be called unless ByteBuffer.arrayOffset() is used or if the ByteBuffer was initialized using ByteBuffer.wrap() or ByteBuffer.allocate().
                                    byte[] data = attachment.array();
                                                                  ^
    (see https://errorprone.info/bugpattern/ByteBufferBackingArray)"

I believe there is a way to just suppress the warning though. So I'll figure that out.

jpd236 replied:
One option would be to just pass Void/null for the attachment parameter and access buffer directly - that would prove to error prone that it's actually the same object which came from ByteBuffer.allocate. As it stands, ErrorProne has no way of knowing this.

sphill99 replied:
Yeah,  I think that should work

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,147 @@
+
+                            @Override
+                            public void failed(Throwable exc, ByteBuffer attachment) {
+                                VolleyLog.d("%s: %s", file.getAbsolutePath(), exc.toString());
```
read failures are abnormal - use e, and log the full stack trace:

VolleyLog.e(exc, "Failed to read file %s", file.getAbsolutePath());

joebowbeer commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,147 @@
+        final File file = getFileForKey(key);
+        final int size = (int) file.length();
+        Path path = Paths.get(file.getPath());
+        AsynchronousFileChannel afc = null;
```
I would move this inside the outer try, which should be a try-catch(IOException), and open a try-finally to close the afc on the next line:

```
try {
  final AsynchronousFileChannel afc = open()
  try {
    // I/O
  } finally {
    afc.close()
  }
} catch (IOException e) {
  // handle e
}
```

See https://github.com/google/volley/blob/master/src/main/java/com/android/volley/toolbox/DiskBasedCache.java#L148

sphill99 replied:
Since the class now requires a higher android api, I switched it to use a try-with-resources. I think this should allow it to just be a simple try-catch, and gets rid of these issues.

sphill99 commented on src/main/java/com/android/volley/AsyncCache.java:
```diff
@@ -0,0 +1,97 @@
+    @Override
+    public final Entry get(String key) {
+        final CountDownLatch latch = new CountDownLatch(1);
+        final AtomicReference<Entry> entryRef = new AtomicReference<>();
```
Ended up not having AsyncCache implement Cache at all, so there was no long a need for this.

sphill99 commented:
I've addressed all the comments and concerns. I updated the try-try finally-catch block, into a try-with-resources, now that we are requiring that SDK. I also made AsyncCache not implement Cache anymore.

jpd236 commented on src/main/java/com/android/volley/AsyncCache.java:
```diff
@@ -0,0 +1,42 @@
+import androidx.annotation.Nullable;
+
+/**
+ * AsyncCache is an abstract class which implements the Cache interface for backwards compatibility,
```
nit: update Javadoc now that we no longer implement Cache. My suggestion would just be:

/** Asynchronous equivalent to the {@link Cache} interface. */

As that is a simple summary of what this class is, and links to Cache in a more structured form so you can leverage the existing docs there if someone is looking for more detail.

jpd236 commented on src/main/java/com/android/volley/AsyncCache.java:
```diff
@@ -0,0 +1,42 @@
+public abstract class AsyncCache {
+
+    public interface OnGetCompleteCallback {
+        void onGetComplete(@Nullable Cache.Entry entry);
```
nit: Could you add a Javadoc for this method? It doesn't have to be super-detailed, but one thing I think would be good to include is when/why entry would be null, and what we do in the event of an error. Something like:

```java
/**
 * Invoked when the read from the cache is complete.
 *
 * @param entry The entry read from the cache, or null if the cache did not contain an entry for the given key or
 *         if the entry couldn't be read.
 */
```

jpd236 commented on src/main/java/com/android/volley/AsyncCache.java:
```diff
@@ -0,0 +1,42 @@
+    public abstract void get(String key, OnGetCompleteCallback callback);
+
+    public interface OnPutCompleteCallback {
+        void onPutComplete(boolean success);
```
Cache#put returns void today, so the introduction of a success boolean is a new concept that I'm not sure we should introduce. Since the focus here is get, I might suggest leaving this out for now.

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,105 @@
+    /** Returns the cache entry with the specified key if it exists, null otherwise. */
+    @Override
+    public void get(String key, OnGetCompleteCallback callback) {
+        final OnGetCompleteCallback cb = callback;
```
I think this line of code is only needed to mark this as final so the inner class can use it? If so, you can just mark the callback param as final (this doesn't require changes to AsyncCache) and use that directly.

sphill99 replied:
yep, I didn't realize I could do that

jpd236 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,105 @@
+                        }
+                    });
+        } catch (IOException e) {
+            VolleyLog.e("%s %s", file.getAbsolutePath(), e.toString());
```
VolleyLog.e(e, "Failed to read file %s", file.getAbsolutePath())

sphill99 commented:
DiskBasedCache and DiskBasedCache tests now uses the shared classes. I moved some stuff (like CountingInputStream) back from DiskBasedCacheUtility to DiskBasedCache, since the async version does not use it. The tests were also refactored and test the code in its new location.

All comments should have been addressed now.

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedCacheUtility.java:
```diff
@@ -0,0 +1,143 @@
+    @VisibleForTesting static final float HYSTERESIS_FACTOR = 0.9f;
+
+    /** Magic number for current version of cache file format. */
+    static final int CACHE_MAGIC = 0x20150306;
```
Note that I don't see this being used outside of CacheHeader, so you could consider removing it from here and keep the one in CacheHeader

sphill99 replied:
Yeah, that's smart. I'll make that change.

jbir789 commented on src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    /** Returns a file object for the given cache key. */
+    public File getFileForKey(String key) {
```
Should this be private?

sphill99 replied:
yep, fixed

joebowbeer commented:
> DiskBasedCache and DiskBasedCache tests now uses the shared classes. I moved some stuff (like CountingInputStream) back from DiskBasedCacheUtility to DiskBasedCache, since the async version does not use it. The tests were also refactored and test the code in its new location.

Thanks!

sphill99 commented on src/main/java/com/android/volley/toolbox/DiskBasedCache.java:
```diff
@@ -374,14 +369,21 @@ private void removeEntry(String key) {
         }
     }
 
+    InputStream createInputStream(File file) throws FileNotFoundException {
```
I had moved them outside of DiskBasedCache, but now since they're there again, I'll add it back

jpd236 replied:
nit: it would lower the diff output if you moved them back to the same location they were in before, which helps in reviewing

jpd236 commented on src/test/java/com/android/volley/toolbox/DiskBasedCacheTest.java:
```diff
@@ -501,86 +503,87 @@ public void testCountingInputStreamByteCount() throws IOException {
         ByteArrayInputStream empty = new ByteArrayInputStream(new byte[] {});
         exception.expect(EOFException.class);
-        DiskBasedCache.readInt(empty);
+        DiskBasedCacheUtility.readInt(empty);
```
For tests which solely exercise methods in DiskBasedCacheUtility, we should move their tests to a new DiskBasedCacheUtilityTest rather than having them in here, since they aren't specific to DiskBasedCache any more.

sphill99 replied:
Fixed this

sophiahaim replied:
Scott, did you upload a new DiskBasedCacheUtilityTest class for the tests you removed from here? I don't see one currently

sophiahaim commented on src/main/java/com/android/volley/AsyncCache.java:
```diff
@@ -0,0 +1,27 @@
+
+    /**
+     * Retrieves an entry from the cache and send it back through the callback function
+     *
```
nit: "sends it back"

Also would be good to link to the callback – {@link OnGetCompleteCallback#onGetComplete}

sphill99 replied:
changed this. I realized I never added the DiskBasedCacheUtilityTest class to the commit, since it was untracked. Should have just fixed that as well.

sphill99 commented:
Uploaded the new test class and fixed the javadocs mistakes.

joebowbeer commented:
By the way, there are 100+ warnings in the Travis build mostly/all about annotations:

https://travis-ci.org/github/google/volley/builds/701650084

I wish the Travis build (or GitHub Action?) reported code coverage.

jpd236 commented:
Filed https://github.com/google/volley/issues/345 for the warnings - these are unrelated to this PR and not a major problem since they only impact the Javadoc (if at all). Coverage is a nice to have, but I don't see it as super critical since we can watch for it in PRs; the project is not that big and doesn't have many people approving changes.
